### PR TITLE
Split WatchlistFinder into 3 query objects and use it

### PR DESCRIPTION
**Pull request by dmarcoux**
It's not possible to sort records from a polymorphic association, since they are fetched in a separate query when calling `includes`. Having a single query by joining with `joins` isn't possible for polymorphic associations (see below). This forces us to start the other way around with our queries, relying directly on the models from the polymorphic association `watchable`. This is why we end up with three query objects.

The error preventing `joins` on polymorphic associations:
```
ActiveRecord::EagerLoadPolymorphicError: Cannot eagerly load the polymorphic association :watchable
```

It's possible to test the changes in the review app:
- API: http://obs-reviewlab.opensuse.org/dmarcoux-watchlist-finders-order/apidocs-new/index.html#/Person/get_person__login__watchlist
- Web UI: Add a few projects/packages/requests to your watchlist

### CODE REVIEW

**Comment by dmarcoux** [comment]
[Add sorting in the API results (M)](https://trello.com/c/XSlBwui3/1886-add-sorting-in-the-api-results-m)

File: src/api/app/queries/packages_for_watchlist_finder.rb [hunk 1]
```diff
@@ -0,0 +1,10 @@
+class PackagesForWatchlistFinder
+  def initialize(relation = Package.all)
```
Comment (saraycp) [inline src/api/app/queries/packages_for_watchlist_finder.rb:2]:
```suggestion
  def initialize(relation = Package.joins(:watched_items))
```

In this watchlist finder, it makes little sense to retrieve the whole list of packages. That's why I would start with a relation that already retrieves packages inside some watchlist.

I'd do the same with all the finders.

File: src/api/app/queries/projects_for_watchlist_finder.rb [hunk 2]
```diff
@@ -0,0 +1,9 @@
+class ProjectsForWatchlistFinder
+  def initialize(relation = Project.all)
```
Comment (saraycp) [inline src/api/app/queries/projects_for_watchlist_finder.rb:2]:
```suggestion
  def initialize(relation = Project.all.joins(:watched_items))
```

File: src/api/app/queries/packages_for_watchlist_finder.rb [hunk 3]
```diff
@@ -0,0 +1,10 @@
...
+    @relation = relation
+  end
+
+  def call(user)
+    @relation.joins(:project).joins(:watched_items).where(watched_items: { user: user })
```
Comment (saraycp) [inline src/api/app/queries/packages_for_watchlist_finder.rb:7]:
```suggestion
    @relation.joins(:project).where(watched_items: { user: user })
```

File: src/api/app/queries/projects_for_watchlist_finder.rb [hunk 4]
```diff
@@ -0,0 +1,9 @@
...
+    @relation = relation
+  end
+
+  def call(user)
+    @relation.joins(:watched_items).where(watched_items: { user: user }).order('LOWER(projects.name), projects.name')
```
Comment (saraycp) [inline src/api/app/queries/projects_for_watchlist_finder.rb:7]:
```suggestion
    @relation.where(watched_items: { user: user }).order('LOWER(projects.name), projects.name')
```

File: src/api/app/queries/requests_for_watchlist_finder.rb [hunk 5]
```diff
@@ -0,0 +1,9 @@
+class RequestsForWatchlistFinder
+  def initialize(relation = BsRequest.all)
+    @relation = relation
+  end
+
+  def call(user)
+    @relation.joins(:watched_items).where(watched_items: { user: user }).order('number DESC')
```
Comment (saraycp) [inline src/api/app/queries/requests_for_watchlist_finder.rb:7]:
```suggestion
    @relation.where(watched_items: { user: user }).order('number DESC')
```

File: src/api/app/queries/requests_for_watchlist_finder.rb [hunk 6]
```diff
@@ -0,0 +1,9 @@
+class RequestsForWatchlistFinder
+  def initialize(relation = BsRequest.all)
```
Comment (saraycp) [inline src/api/app/queries/requests_for_watchlist_finder.rb:2]:
```suggestion
  def initialize(relation = BsRequest.all.joins(:watched_items))
```

**Review by saraycp — CHANGES_REQUESTED** [review]

File: src/api/app/queries/packages_for_watchlist_finder.rb (hunk 1 above)
Comment (dmarcoux) [reply src/api/app/queries/packages_for_watchlist_finder.rb:2]: Sure :+1:

**Review by saraycp — APPROVED** [review]

**Pull request merged**